### ISSUE TITLE
Confirm if a partition is needed/relevant for a Full Sync?

### ISSUE THREAD

**Issue by JayVDZ** [TetronIO/JIM#72]
## Summary

Currently, when creating a run profile, the UI requires you to select a partition (for connectors that support partitions, such as LDAP). However, the selected partition is **stored but completely ignored** during sync execution. This means:

1. The UI enforces partition selection, but integration tests create run profiles without partitions and work fine
2. Connectors query **all selected partitions** regardless of the run profile's partition setting
3. The partition field on `ConnectedSystemRunProfile` is effectively dead data

This issue tracks implementing partition filtering for run profiles, starting with import.

### UI Layer
- Partition dropdown appears when `ConnectedSystem.ConnectorDefinition.SupportsPartitions` is `true`
- When shown, partition selection is marked as `Required` via MudBlazor validation
- Only partitions marked as `Selected` are shown in the dropdown
- Page Size is shown unconditionally for all run profile types, including sync — but sync processors ignore it entirely (they use a system-wide setting from `GetSyncPageSizeAsync()`)
- **File:** `src/JIM.Web/Pages/Admin/Components/ConnectedSystemRunProfilesTab.razor` (lines 138–149, 196–207)

### API Layer
- `PartitionId` is **nullable** in both `CreateRunProfileRequest` and `UpdateRunProfileRequest` DTOs
- The API happily accepts run profiles without a partition, even for connectors that support partitions
- No server-side validation enforces partition requirement
- **File:** `src/JIM.Web/Controllers/Api/SynchronisationController.cs` (lines 1224–1232)

### Business Logic Layer
- `IsRunProfileValid()` only validates run type support, **not** whether a partition is set
- **File:** `src/JIM.Application/Servers/ConnectedSystemServer.cs` (lines 3868–3883)

### Connector Layer (Import) — The Core Problem
- The LDAP connector iterates over **all selected partitions** from `_connectedSystem.Partitions.Where(p => p.Selected)`, completely ignoring `_runProfile.Partition`
- This means every import run profile queries every selected partition, regardless of the partition assigned to that run profile
- **File:** `src/JIM.Connectors/LDAP/LdapConnectorImport.cs` (lines 95–120 for full import, line 180 for delta import)

### Sync Processors
- Full Sync and Delta Sync processors load CSOs without any partition filtering
- Sync processors use a system-wide page size from `GetSyncPageSizeAsync()`, completely ignoring the run profile's `PageSize` property

### Integration Tests
- All integration test scenarios in `Setup-Scenario1.ps1` create run profiles **without** specifying partitions — even for LDAP, which has partitions imported and selected
- Tests pass because no layer below the UI enforces partition requirement

### Model
- `ConnectedSystemRunProfile.Partition` is `ConnectedSystemPartition?` (nullable)
- The partition ID is persisted to the database (FK relationship exists) but never read during execution
- **File:** `src/JIM.Models/Staging/ConnectedSystemRunProfile.cs` (lines 62–64)

## Gap Summary

| Layer | Behaviour |
|-------|-----------|
| **UI** | Partition required when `SupportsPartitions=true` (all run types); Page Size shown for all run types |
| **API** | Partition optional (nullable) |
| **Business logic validation** | No partition validation |
| **Connectors (import)** | Ignore run profile partition; query ALL selected partitions |
| **Connectors (export)** | Ignore run profile partition |
| **Sync processors (full/delta)** | No partition filtering; ignore run profile PageSize |
| **Integration tests** | Don't set partition at all |

## Consequences

1. **Sub-optimal queries:** For LDAP with multiple selected partitions, every import run profile queries all selected partitions rather than just the targeted one
2. **No granular control:** Users cannot schedule different import cadences per partition (e.g., import from a high-change partition hourly but a stable partition daily)
3. **Misleading UI:** The UI collects partition selection and marks it as required, giving users the impression their choice matters — but it has no effect. Similarly, the Page Size field is shown for sync run profiles where it is ignored.

### Phase 1: Import Partition Scoping ✅

Partition scoping for import run profiles is a complete, standalone feature that delivers value without any dependency on later phases:

- **Query efficiency** — Instead of querying every selected partition (potentially across multiple domain controllers in different sites/regions), only the targeted partition is queried. For large directories with many partitions, this significantly reduces load on both JIM and the directory infrastructure.
- **Granular scheduling** — A high-churn partition (e.g., a contractor OU that changes daily) can be imported hourly, while a stable partition (e.g., service accounts) imports weekly.
- **Operational control** — If one partition's domain controller is under maintenance, imports from other partitions are unaffected. If an import from one partition is failing, it doesn't block the rest.

Import is the **ingest** stage — it controls what enters the staging area. Sync and export operate on whatever is already staged, so they are unaffected by how objects were imported. Partition-scoped import simply gives administrators control over *when* and *what* they pull in, without creating any partial-support issues downstream.

### Phase 1b: Multi-Partition Integration Tests with OpenLDAP ✅

Branch: `feature/openldap-integration-testing` | Plan: `docs/plans/done/OPENLDAP_INTEGRATION_TESTING.md`

OpenLDAP introduced as a second LDAP backend for integration testing. All 6 phases complete — all scenarios pass against both SambaAD and OpenLDAP at Medium scale.

**OpenLDAP Docker infrastructure (Phase 1):** ✅
- Configure OpenLDAP service with multiple suffixes in `docker-compose.integration-tests.yml` (`dc=yellowstone,dc=local` + `dc=glitterband,dc=local`)
- Dockerfile based on `bitnamilegacy/openldap` (OpenLDAP 2.6.10) with init script for second MDB database
- Build script (`Build-OpenLdapImage.ps1`) with content-hash labelling
- Accesslog overlay enabled for delta import testing

**Test data population (Phase 2):** ✅
- `Populate-OpenLDAP.ps1` creates distinct `inetOrgPerson` users and `groupOfNames` groups per suffix
- Verified at Nano and Micro scales

**Test framework parameterisation (Phase 3):** ✅
- `Get-DirectoryConfig` abstraction in `Test-Helpers.ps1` (SambaAD + OpenLDAP configs)
- `-DirectoryType` parameter on `Run-IntegrationTests.ps1` with interactive menu
- Refactored `LDAP-Helpers.ps1`, `Setup-Scenario1.ps1`, `Invoke-Scenario1.ps1` to accept `$DirectoryConfig`

**Connector fixes (Phase 4):** ✅
- `LdapDirectoryType` enum (`ActiveDirectory`, `SambaAD`, `OpenLDAP`, `Generic`) — directory-specific behaviour centralised in `LdapConnectorRootDse`
- RFC 4512 schema discovery via `cn=Subschema` (`Rfc4512SchemaParser.cs`) with 37 unit tests
- Partition discovery via `namingContexts` rootDSE attribute for non-AD directories
- `entryUUID` recommended as external ID for non-AD directories
- `distinguishedName` synthesised during import (OpenLDAP doesn't return it as an attribute)
- Accesslog-based delta import (`GetDeltaResultsUsingAccesslog`) using `reqStart` timestamp watermarks
- OpenLDAP detection via `structuralObjectClass: OpenLDAProotDSE`
- DN-aware RDN attribute detection in export (`IsRdnAttribute` parses RDN from DN)
- "Include Auxiliary Classes" connected system setting for schema discovery

**End-to-end validation (Phase 5):** ✅
- Scenario 1 (HR → OpenLDAP) passing at Nano scale — all 8 test steps
- Accesslog-based delta import confirmed working for export confirmation
- AD-specific tests (Disable/Enable) gracefully skipped for OpenLDAP
- Mover Rename verifies `cn` update (not DN change) for OpenLDAP (uid-based RDN)
- Mover Move verifies `departmentNumber` update (not OU move) for OpenLDAP (flat OU structure)

**Phase 6 — Scenarios (complete):** ✅
- ✅ S9 (Partition-Scoped Imports) — true multi-partition filtering with Yellowstone + Glitterband
- ✅ S7 (Clear CS Objects) — DirectoryConfig threading
- ✅ S6 (Scheduler Service) — system name + docker cp fix
- ✅ S2 (Cross-Domain Sync) — all 4 tests passing (unblocked by #435)
- ✅ S5 (Matching Rules) — DirectoryConfig threading, user cleanup parameterised
- ✅ S4 (Deletion Rules) — all 7 tests passing, LDAP-Helpers replace samba-tool
- ✅ S8 (Cross-Domain Entitlement Sync) — all 6 tests passing at Medium. Accesslog mapsize/sizelimit configured. Delta import null watermark fix. Connector warnings moved to Activity.WarningMessage.
- ⏭ S3 (GAL Sync) — deferred, out of scope for this phase

**Phase 6 — Infrastructure (complete):** ✅
- ✅ OpenLDAP Docker image snapshotting — `Build-OpenLDAPSnapshots.ps1` with content-hash staleness detection
- ✅ Scenario 9 updated for true multi-partition filtering assertions

**Phase 6 — Verification (complete):** ✅
- ✅ Samba AD full regression — 8/8 scenarios passed (Small template, 2026-04-01)
- ✅ OpenLDAP full regression — 8/8 scenarios passed (Small template, 2026-04-01)
- ✅ `-DirectoryType All` test runner option — runs both directory types sequentially

**Phase 6 — Scale Testing (complete):** ✅
- ✅ Full regression at Medium on both SambaAD and OpenLDAP (8/8 scenarios each, 2026-04-01)
- ✅ Fixed: Two-phase parallel MVO persistence to prevent cross-partition FK violations at Medium+ scale

### Phase 2: Evaluate Sync Partition Scoping (future)

Evaluate whether and how partition-scoped synchronisation could be supported. Key challenges to investigate:

- `ConnectedSystemObject` (CSO) **does not currently store which partition it was imported from** — there is no `PartitionId` FK on the CSO model. A schema migration would be required to add this.
- Sync needs a **holistic view of all CSOs** to correctly resolve joins and avoid duplicate projections. Restricting sync to a single partition's CSOs could cause missed joins or incorrect projections if objects in other partitions are relevant.
- Need to determine whether partition-scoped sync has real-world value or whether the holistic approach is always correct.

### Phase 3: Evaluate Export Partition Scoping (future)

Evaluate whether and how partition-scoped export could be supported. This carries the most design complexity:

- Requires `PartitionId` on CSOs (same prerequisite as Phase 2) so pending exports can be routed to the correct partition.
- **Cross-partition reference dependencies** are a significant risk. Consider: users in Partition A, groups in Partition B. If an export is scoped to Partition B and JIM needs to both provision a new user in Partition A and add that user as a group member in Partition B, the membership export would fail — the referenced user doesn't exist yet because its partition wasn't included in the export.
- Solving this would require detecting cross-partition references in pending exports, building a dependency graph across partitions, and either pulling in dependent exports (defeating the purpose of scoping) or deferring/failing gracefully.
- Need to evaluate whether this complexity is justified and whether cross-partition references are common in real-world deployments.

## Design Decision: Partition is Optional on Import Run Profiles

Setting a partition on an import run profile is **optional by design**. This gives administrators the choice:
- **With partition:** Scoped import of a specific partition — more efficient, enables granular scheduling
- **Without partition:** Import across all selected partitions — simpler when partition-level control isn't needed

### Phase 1 — Complete ✅

**Connectors (import):**
- ✅ LDAP connector respects `runProfile.Partition` when set via `GetTargetPartitions()` helper
- ✅ Falls back to all selected partitions when partition is null
- ✅ Both Full Import and Delta Import paths updated

**UI:**
- ✅ Partition dropdown only shown for import run profile types (Full Import, Delta Import)
- ✅ Partition is optional with `Clearable="true"` and helper text ("Optional. Leave empty to import from all selected partitions.")
- ✅ Page Size only shown for import run profile types
- ✅ Connector detail page capabilities grouped by category (Import, Export, Hierarchy, Data Handling)

**Integration tests (Scenario 9):**
- ✅ Setup script creates LDAP connected system with partition and container selection
- ✅ Three run profiles: scoped import, unscoped import, and full sync
- ✅ Validates scoped import creates CSOs from the targeted partition
- ✅ Validates unscoped import completes successfully
- ✅ Validates metaverse consistency (Full Sync) after both import paths
- ✅ Registered in test runner, passing on Micro template

### Phase 1b — Phases 1–6 Complete ✅

See plan: [`docs/plans/done/OPENLDAP_INTEGRATION_TESTING.md`](docs/plans/done/OPENLDAP_INTEGRATION_TESTING.md)

**Complete:**
- ✅ OpenLDAP Docker infrastructure with two suffixes and accesslog overlay
- ✅ Test data population (`Populate-OpenLDAP.ps1`) — Nano and Micro scales
- ✅ Test framework parameterisation (`-DirectoryType` parameter, `Get-DirectoryConfig`)
- ✅ LDAP connector fixes: `LdapDirectoryType` enum, RFC 4512 schema discovery (37 unit tests), `entryUUID` support, `distinguishedName` synthesis, accesslog delta import, DN-aware RDN detection
- ✅ Scenario 1 passing against OpenLDAP — all 8 test steps (Joiner, Mover x3, Disable/Enable skipped, Leaver, Reconnection)
- ✅ All scenarios (S1-S9, excluding S3 deferred) parameterised and passing on both SambaAD and OpenLDAP
- ✅ Scale testing — full regression at Medium on both directory types (2026-04-01)
- ✅ Fixed: Two-phase parallel MVO persistence to prevent cross-partition FK violations at Medium+ scale
